### ISSUE TITLE
Gradient of the log probability

### ISSUE THREAD

**Issue by arrjon** [bayesflow-org/bayesflow#244]
It would be nice to be able to compute the gradient of the log probability.

For example, with the jax-backend, I would like to be able to do something along this lines:
```
from jax import grad

def partial_objective(theta):
    log_prop = approximator.log_prob(
        data={'theta': theta, 'x': x},
        batch_size=1,
    )
    return log_prop[0]

grad(partial_objective)(x)
```

The main reason this is not working at the moment, is the adapter. Any idea, how this could be done?

**Comment by stefanradev93** [comment]
Agreed that this is an important feature to have! We may want to implement an internal function `log_prob_and_grad` that does not trace the adapter transformations, but the naive implementation would give gradients wrt the *transformed* inputs, which may not be what one wants (e.g., gradients would also have a different scale).

**Comment by LarsKue** [comment]
I will look into what options we have to allow something like this. For now, you can use the protected `_log_prob` as a work-around. This returns a `keras.backend.Tensor` so the gradient should be enabled. The complete equivalent code should be:

```py3
def partial_objective(theta):
    log_prob = approximator._log_prob(**approximator.adapter(data={"theta": theta, "x": x}))
    return log_prob[0]
```

**Comment by LarsKue** [comment]
We might implement this directly with a `score` method.

**Comment by stefanradev93** [comment]
@arrjon Should we add this to our inference nets?

**Comment by arrjon** [comment]
Computing the gradient of the log probability would be nice, e.g. so one can use it in a MCMC sampler when the inference network approximates a likelihood. But not sure, how often this will be actually used.

**Comment by paul-buerkner** [comment]
I think for our self-consistency losses, we have already needed (and used) the gradient of log prob. @Kucharssim and @aayush2505 can you provide some insights?
